### PR TITLE
Replace `ParamEnv` with a new type in chalk context.

### PR DESCRIPTION
**Pull request by scalexm**
I left a few FIXMEs.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_traits/chalk_context.rs [hunk 1]
```diff
@@ -97,24 +99,24 @@ impl context::Context for ChalkArenas<'tcx> {
 
     type Parameter = Kind<'tcx>;
 
-    type ProgramClause = ProgramClause<'tcx>;
```
Comment (tmandry) [inline src/librustc_traits/chalk_context.rs]: Why doesn’t the terminology line up here anymore with chalk?

File: src/librustc_traits/lowering/environment.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    // The environment of an impl Trait type is its defining function's environment.
+    if let Some(parent) = ty::is_impl_trait_defn(tcx, def_id) {
+        return environment(tcx, parent);
+    }
+
+    // Compute the bounds on `Self` and the type parameters.
+    let ty::InstantiatedPredicates { predicates } =
+        tcx.predicates_of(def_id).instantiate_identity(tcx);
+
+    let clauses = predicates.into_iter()
```
Comment (tmandry) [inline src/librustc_traits/lowering/environment.rs]: I love how clean this is.

File: src/librustc_traits/chalk_context.rs (hunk 1 above)
Comment (scalexm) [reply src/librustc_traits/chalk_context.rs]: Mmh at some point the terminology diverged, we use `Clause` in rustc for `ProgramClause` in chalk, and `ProgramClause` in rustc for `ProgramClauseImplication` in chalk. I agree we should somehow resolve that conflict.

Comment (nikomatsakis) [reply src/librustc_traits/chalk_context.rs]: I think the meaning of these words I would **like** to be using:

- "clause" is one of the "base facts" we know to be true
    - in particular, this includes clauses injected via hypothesis, e.g. the clause `A` in `A => G`.
- "program clause" is something that was given in the "input program"
    - in our case, this means the stuff we created by lowering impls and so forth

In particular, program clauses are available globally at all times, whereas clauses may only be available to some goals.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me once rebased

**Pull request merged**